### PR TITLE
Closes #357 -  Create PointsBalanceEvent with appropiate eventListener

### LINKED ISSUE #357: Create PointsBalanceEvent with appropiate eventListener

### PR DESCRIPTION
**Pull request by volodymyrglynka**

### CODE REVIEW

**Review by OleksandrMazur — APPROVED** [review]

**Review by maxturko — CHANGES_REQUESTED** [review]
There is an error in Validators test, your version of project is failing to build

Tests in error:
  UserControllerTest.testGetAuthorizedUserPublicInfo:49 ? NestedServlet Request ...
 please, resolve

File: src/main/java/com/softserve/academy/spaced/repetition/config/PointsBalanceEvent.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    public PointsBalanceEvent(Object source, User user) {
+        super(source);
+        this.user = user;
+    }
+
+    public User getUser() {
+        return user;
+    }
+
+    public void setUser(User user) {
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/config/PointsBalanceEvent.java]: Please remove the setter to make User required.

File: src/main/java/com/softserve/academy/spaced/repetition/config/EventListenerBean.java [hunk 2]
```diff
@@ -38,6 +38,13 @@ public void handleCheckPointsEvent(ContextRefreshedEvent event) {
             }
         }
     }
+
+    @EventListener
+    public User updatePointsBalance(PointsBalanceEvent pointBalanceEvent) {
+        LOG.info("Updating points balance for {} {}", pointBalanceEvent.getUser().getPerson().getFirstName(),
+                pointBalanceEvent.getUser().getPerson().getLastName());
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/config/EventListenerBean.java]: A possible NullPointerException. I would write 
`LOG.info("Updating points balance for {}.", pointBalanceEvent.getUser())`
See my comment about making User required.

File: src/main/java/com/softserve/academy/spaced/repetition/controller/UserController.java [hunk 3]
```diff
@@ -34,12 +36,15 @@
...
+    @Autowired
+    private ApplicationEventPublisher publisher;
+
     @GetMapping("api/user/details")
     @PreAuthorize("isAuthenticated()")
     public ResponseEntity<UserDTO> getAuthorizedUserPublicInfo() throws NotAuthorisedUserException {
         User user = userService.getAuthorizedUser();
         if (!userService.isAdmin(user)) {
-            userService.updatePointsBalance(user);
+            publisher.publishEvent(new PointsBalanceEvent(this, user));
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/UserController.java]: Please do not pass `this`. `this` is a managed Spring bean with different dependencies and access to different methods, that an event shouldn't access.

File: src/main/java/com/softserve/academy/spaced/repetition/service/impl/PointsTransactionServiceImpl.java [hunk 4]
```diff
@@ -62,8 +68,8 @@ public void buyDeck(Long deckId) throws NotAuthorisedUserException, PointsTransa
             transaction.setCreationDate(new Date());
             transaction.setReference(deckOwnership.getReference());
             transactionRepository.save(transaction);
-            userService.updatePointsBalance(userFrom);
-            userService.updatePointsBalance(userTo);
+            publisher.publishEvent(new PointsBalanceEvent(this, userFrom));
```
Comment (antonbakalets) [inline src/main/java/com/softserve/academy/spaced/repetition/service/impl/PointsTransactionServiceImpl.java]: The same here. For example:
`this.getClass().getCanonicalName()`
to know the source of the evnt.

**Review by antonbakalets — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/academy/spaced/repetition/config/PointsBalanceEvent.java (hunk 1 above)
Comment (volodymyrglynka) [reply src/main/java/com/softserve/academy/spaced/repetition/config/PointsBalanceEvent.java]: Thanks for comments. You are right, there was an error in test. I fixed it and removed setter from event.

File: src/main/java/com/softserve/academy/spaced/repetition/service/impl/PointsTransactionServiceImpl.java (hunk 4 above)
Comment (volodymyrglynka) [reply src/main/java/com/softserve/academy/spaced/repetition/service/impl/PointsTransactionServiceImpl.java]: Thanks for remarks. I made changes.

**Review by OleksandrBartosh — APPROVED** [review]

File: src/test/java/com/softserve/academy/spaced/repetition/controller/UserControllerTest.java [hunk 5]
```diff
@@ -46,6 +52,7 @@ public void setUp() {
     @Test
     public void testGetAuthorizedUserPublicInfo() throws Exception {
         when(userService.getAuthorizedUser()).thenReturn(getUser());
+        doNothing().when(publisher).publishEvent(new PointsBalanceEvent(this.getClass().getCanonicalName() , getUser()));
```
Comment (antonbakalets) [inline src/test/java/com/softserve/academy/spaced/repetition/controller/UserControllerTest.java]: This should be
`doNothing().when(publisher).publishEvent(Matchers.any(PointsBalanceEvent.class));`

**Review by antonbakalets — APPROVED** [review]

File: src/test/java/com/softserve/academy/spaced/repetition/controller/UserControllerTest.java (hunk 5 above)
Comment (volodymyrglynka) [reply src/test/java/com/softserve/academy/spaced/repetition/controller/UserControllerTest.java]: Done

**Review by maxturko — APPROVED** [review]

**Pull request merged**